### ISSUE TITLE
numpy ufuncs don't have __qualname__

### ISSUE THREAD

**Issue by pitrou** [numpy/numpy#4952]
Numpy ufuncs have a `__name__`, but don't have a `__qualname__` as defined in PEP 3155. Similarly, they don't seem to define a `__module__`.
(note that for numpy's builtin ufuncs, the `__qualname__` should simply be the same as the `__name__`, e.g. `np.sin.__qualname__ == 'sin'`)

**Comment by juliantaylor** [comment]
I'm going to close this, I'm against adding it in my interpretation of the pep it only applies to types and not objects.  But a statement from a python dev clarifying the scope of the pep can convince me otherwise.

**Comment by pitrou** [comment]
Well, I'm the author of the PEP.

**Comment by juliantaylor** [comment]
interesting to know, can you then say why you want objects to have qualname? the pep is not clear about it.
e.g. python itself does not give objects qualnames

**Comment by juliantaylor** [comment]
not even the one most similar to a numpy ufunc, namedtuple:

```
Point = collections.namedtuple('Point', ['x', 'y'])
Point(11, y=22).__qualname__
AttributeError: 'Point' object has no attribute '__qualname__'
```

**Comment by pitrou** [comment]
My general intuition is that a ufunc is in many contexts used like a regular function, so like a regular function it would be reasonable for it to have a `__qualname__`. My other reason is that it already has a `__name__`, which also mimicks functions in that regard.

**Comment by juliantaylor** [comment]
does python provide a way to declare an instance of a ufunc is a function so it gets qualname automatically?
Though I still don't see enough advantage of adding it that is worth the time discussing it further.

**Comment by juliantaylor** [comment]
please treat my comments as +0 if someone else of the devs feels like this is worthwhile feel free to reopen and merge the pr. I don't want to spend any more time on it.
Also please note my opinion is purely about the need of numpy to add this, the pep itself is very useful for python3, thank you for doing that.

**Comment by pitrou** [comment]
> does python provide a way to declare an instance of a ufunc is a function so it gets qualname automatically?

No. For pure Python functions, Python computes `__qualname__` at compile time. For obvious reasons that can't work for ufuncs (they aren't declared using standard ``def`` syntax).
However, Python does add a `__qualname__` to non-Python functions, e.g.:

```
>>> len.__name__
'len'
>>> len.__qualname__
'len'
```

The basic point is to make all function-like objects similar in exposed metadata, to ease the job of 3rd-party code.

**Comment by charris** [comment]
@juliantaylor Is there a good reason *not* to add it?

**Comment by tgy** [comment]
`__qualname__` is still missing from ufunc objects in 2020. Maybe this issue should be re-opened?

**Comment by eric-wieser** [comment]
Agreed, I think we should have this. There are projects that create ufuncs as class methods (numba?), and those don't pickle correctly unless they have the capability to set `__qualname__`.

**Comment by arubiales** [comment]
Hi I'm new and this is my first contribution to numpy. I have read the Guidelines and everythin is clear to me.

I want to work on this issue, but I can't find the module or files to start.

¿Can you help me please?

**Comment by StephenTanksley** [comment]
take

**Comment by seberg** [comment]
@arubiales its defined in C in `numpy/core/src/umath/ufunc_object.c` so you should be comfortable with the C-API if you look at this.  One problem is that we either need to make this attribute mutable from Python, or you need to use the current `__name__` (in a sense). That is similar to static types in C, we allow users to set the `name`, and for purposes of pickling, I use the fact that `__name__` could include a `.`. But if we add `__qualname__` we should probably store only what comes after the last `.` in the `__name__` and store everything else in qualname (that might require modifying the pickle code, I doubt that, but we have to remember to check).

**Comment by seberg** [comment]
> `__name__` attributes don't contain periods, and `__qualname__` only does so in the case of methods i.e. irrelevant in the case of ufuncs

This is not irrelevant, because a ufunc _can_ be attached to a different object (and this can be the only place it lives). Numba does this.  (It is irrelevant for NumPy or SciPy, but not generally.)

**Comment by seberg** [comment]
@mtsokol since you have added `__module__` maybe you can actually also add `__qualname__` for the ones in NumPy (with a test that it matches the name)?

It is really part of the same issue since `__qualname__` is more correct than name.

**Comment by mtsokol** [comment]
Solved in #27877.
